realm/src/androidTest/java/io/realm/RealmInMemoryTest.java

```diff
@@ -0,0 +1,242 @@
+                .build();
+
+        Realm.deleteRealm(conf);
+        Realm.deleteRealm(encConf);
+
+        testRealm.beginTransaction();
+        Dog dog = testRealm.createObject(Dog.class);
+        dog.setName("DinoDog");
+        testRealm.commitTransaction();
+
+        // Test a normal Realm file
+        testRealm.writeCopyTo(new File(getContext().getFilesDir(), fileName));
+        Realm onDiskRealm = Realm.getInstance(conf);
+        assertEquals(onDiskRealm.allObjects(Dog.class).size(), 1);
+        onDiskRealm.close();
+
+        // Test a encrypted Realm file
+        testRealm.writeEncryptedCopyTo(new File(getContext().getFilesDir(), encFileName), key);
+        onDiskRealm = Realm.getInstance(encConf);
+        assertEquals(onDiskRealm.allObjects(Dog.class).size(), 1);
+        onDiskRealm.close();
+        // Test with a wrong key to see if it fails as expected.
+        try {
+            RealmConfiguration wrongKeyConf = new RealmConfiguration.Builder(getContext())
+                    .name(encFileName)
+                    .encryptionKey(TestHelper.getRandomKey(42))
+                    .build();
+            Realm.getInstance(wrongKeyConf);
+            fail("Realm.getInstance should fail with illegal argument");
+        } catch (IllegalArgumentException ignored) {
+        }
+    }
+
+    // Test below scenario:
+    // 1. Create a in-memory Realm instance in the main thread.
+    // 2. Create a in-memory Realm with same name in another thread.
+    // 3. Close the in-memory Realm instance in the main thread and the Realm data should not be released since
+    //    another instance is still held by the other thread.
+    // 4. Close the in-memory Realm instance and the Realm data should be released since no more instance with the
+    //    specific name exists.
+    public void testMultiThread() throws InterruptedException, ExecutionException {
+        final AtomicBoolean isWorkerThreadReady = new AtomicBoolean(false);
+        final AtomicBoolean isWorkerThreadFinished = new AtomicBoolean(false);
+        final AtomicBoolean isRealmInMainThreadClosed = new AtomicBoolean(false);
+
+        // Step 2.
+        Thread workerThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Realm realm = Realm.getInstance(inMemConf);
+                realm.beginTransaction();
+                Dog dog = realm.createObject(Dog.class);
+                dog.setName("DinoDog");
+                realm.commitTransaction();
+                isWorkerThreadReady.set(true);
+
+                assertEquals(realm.allObjects(Dog.class).size(), 1);
+
+                // Wait until Realm instance closed in main thread
+                try {
+                    while (!isRealmInMainThreadClosed.get()) {
+                        Thread.sleep(5);
+                    }
+                } catch (InterruptedException e) {
+                    fail("Worker thread was interrupted.");
```

Comment:
This fail will not report in JUnit, you need to capture it & throw it within the testcase's Thread 
ex: https://github.com/realm/realm-java/blob/nh-async-query/realm/src/androidTest/java/io/realm/RealmAsyncQueryTests.java#L241 You can also use `CountDownLatch` to avoid _"busy waiting"_  `while (condition) Thread.sleep();`